### PR TITLE
HW03 Gerasev Georgii

### PR DESCRIPTION
**Pull request by VenInf**
It's not clear, how the "pow" operation should function, since it can be used as root, but doesn't have any potential errors in the description.

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -2,36 +2,113 @@ module Main where
...
 instance Eq Expr where 
-  (==) = undefined 
+  (==) (Num ex1) (Num ex2) = ex1 == ex2
+  (==) (Root ex1) (Root ex2) = ex1 == ex2
+  (==) (Plus a1 a2) (Plus b1 b2) = a1 == b1 && a2 == b2 
+  (==) (Minus a1 a2) (Minus b1 b2) = a1 == b1 && a2 == b2 
+  (==) (Mult a1 a2) (Mult b1 b2) = a1 == b1 && a2 == b2 
+  (==) (Div a1 a2) (Div b1 b2) = a1 == b1 && a2 == b2
+  (==) (Pow a1 a2) (Pow b1 b2) = a1 == b1 && a2 == b2 
+  (==) _ _ = False
```
Comment (kajigor) [inline Main.hs:36]: can be derived

File: Main.hs [hunk 2]
```diff
@@ -2,36 +2,113 @@ module Main where
...
 
-data Error 
+data Error = DivideByZero String
+            | NegativeFromRoot String 
+            | PowNaN String 
+            | ComplexError Error Error
```
Comment (kajigor) [inline Main.hs:40]: :+1:

File: Main.hs [hunk 3]
```diff
@@ -2,36 +2,113 @@ module Main where
...
+  show (PowNaN msg) = "PowNaN error " ++ show msg
+  show (ComplexError error1 error2) = "ComplexErrorOf " ++ show error1 ++ " and " ++ show error2
+  
 
 instance Eq Error where 
-  (==) = undefined 
+  (==) (DivideByZero msg1) (DivideByZero msg2) = msg1 == msg2
+  (==) (NegativeFromRoot msg1) (NegativeFromRoot msg2) = msg1 == msg2 
+  (==) (PowNaN msg1) (PowNaN msg2) = msg1 == msg2 
+  (==) _ _ = False
```
Comment (kajigor) [inline Main.hs:53]: can be derived

File: Main.hs [hunk 4]
```diff
@@ -2,36 +2,113 @@ module Main where
...
+
+andExprError :: (Expr -> Expr -> Expr) -> Either Expr Error -> Either Expr Error -> Either Expr Error -- priority for the errors, operands otherwise
+andExprError _ (Right error1) (Right error2) = Right (ComplexError error1 error2)
+andExprError _ (Left expr1) (Right error2) = Right error2
+andExprError _ (Right error1) (Left expr2) = Right error1
+andExprError oper (Left expr1) (Left expr2) = Left (oper expr1 expr2)
```
Comment (kajigor) [inline Main.hs:59]: Please use `Left` for errors and `Right` for the correct answers. The mnemonic is `Right` for the right answer

Comment (kajigor) [reply Main.hs:59]: It'll become clear why this is important once we learn about monads.

File: Main.hs [hunk 5]
```diff
@@ -2,36 +2,113 @@ module Main where
...
+exprToNum (Num n) = n
 
-eval :: Expr -> Either Error Double 
-eval = undefined 
+eval :: Expr -> Either Error Double
+eval expr = either (Right . exprToNum) Left (resolve (Left expr)) 
 
 cases :: [(Expr, Either Error Double)]
-cases = undefined 
+cases = [
```
Comment (kajigor) [inline Main.hs:97]: Not all cases are covered

**Review by kajigor — CHANGES_REQUESTED** [review]

**Comment by VenInf** [comment]
Fixed all of the commented parts.

**Comment by kajigor** [comment]
ok

**Pull request closed without merging**